guides/micronaut-kafka/books/groovy/src/test/groovy/example/micronaut/BookControllerSpec.groovy

```diff
@@ -0,0 +1,96 @@
+package example.micronaut
+
+import io.micronaut.configuration.kafka.annotation.KafkaListener
+import io.micronaut.configuration.kafka.annotation.Topic
+import io.micronaut.context.ApplicationContext
+import io.micronaut.core.type.Argument
+import io.micronaut.http.HttpRequest
+import io.micronaut.http.client.HttpClient
+import io.micronaut.http.client.exceptions.HttpClientResponseException
+import io.micronaut.runtime.server.EmbeddedServer
+import org.testcontainers.containers.KafkaContainer
+import org.testcontainers.utility.DockerImageName
+import spock.lang.AutoCleanup
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.util.concurrent.PollingConditions
+
+import java.util.concurrent.ConcurrentLinkedDeque
+
+import static io.micronaut.configuration.kafka.annotation.OffsetReset.EARLIEST
+
+class BookControllerSpec extends Specification {
+
+    private static final Collection<Book> received = new ConcurrentLinkedDeque<>();
+    private static final PollingConditions conditions = new PollingConditions(timeout: 5)
+
+    @Shared
+    @AutoCleanup
+    static KafkaContainer kafka = new KafkaContainer(DockerImageName.parse('confluentinc/cp-kafka:latest'))
+
+    @Shared @AutoCleanup EmbeddedServer embeddedServer
+    @Shared @AutoCleanup ApplicationContext context
+    @Shared @AutoCleanup HttpClient client
+
+    void setupSpec() {
+        kafka.start()
+        embeddedServer = ApplicationContext.run(EmbeddedServer,
+                ['kafka.bootstrap.servers': kafka.bootstrapServers])
+        context = embeddedServer.applicationContext
+        client = context.createBean(HttpClient, embeddedServer.URL)
+    }
```

Comment:
is there a reason you didn't use `TestPropertyProvider` here for Spock in the same way?

Reply:
Fixed